### PR TITLE
TextField and TextArea helper text #DS-597

### PR DESCRIPTION
**Pull request by tomassychra**
# Adding helper text for TextField and TextArea [#DS-597](https://jira.lmc.cz/browse/DS-597)

## What was done:
* Web:
   * **TextField**
      * styles for helperText element
      * example
      * docs
   * **TextArea**
      * styles for helperText element
      * example
      * docs
* Twig:
   * **TextField**
      * docs
   * **TextArea**
      * docs
   * **TextFieldBase**
      * props for helper text
      * element for helper text
      * docs
 * React:
    * **TextField**
       * props
       * docs
       * stories
       * tests
    * **TextArea**
       * props
       * docs
       * stories
       * tests
    * **TextFieldBase**
       * element for helper text
       * `useTextFieldBaseStyleProps` update

### CODE REVIEW

**Review by literat — CHANGES_REQUESTED** [review]
A little bit of nitpicking. Otherwise it seems fine. 👌

**Comment by literat** [comment]
HTML Demo: https://deploy-preview-718--spirit-design-system-demo.netlify.app/

File: packages/web/src/scss/components/TextField/index.html [hunk 1]
```diff
@@ -252,6 +252,18 @@
...
+
+  <div class="TextField">
+    <label for="textfieldWithHelperText" class="TextField__label TextField__label--required">Label</label>
+    <input type="text" id="textfieldWithHelperText" class="TextField__input" placeholder="Placeholder" />
+    <div class="TextField__helperText">* The field is required</div>
+  </div>
+  <div class="TextField TextField--danger">
+    <label for="textfieldWithHelperTextAndValidationMessage" class="TextField__label TextField__label--required">Label</label>
+    <input type="text" id="textfieldWithHelperTextAndValidationMessage" class="TextField__input" placeholder="Placeholder" />
+    <div class="TextField__helperText">* The field is required</div>
```
[the remark this answers is not in the document]
Comment (tomassychra) [reply packages/web/src/scss/components/TextField/index.html:264]: done

File: packages/web/src/scss/components/TextField/index.html [hunk 2]
```diff
@@ -252,6 +252,18 @@
     </div>
     <button type="button" class="Button Button--primary Button--medium">Button</button>
   </div>
+
+  <div class="TextField">
+    <label for="textfieldWithHelperText" class="TextField__label TextField__label--required">Label</label>
+    <input type="text" id="textfieldWithHelperText" class="TextField__input" placeholder="Placeholder" />
+    <div class="TextField__helperText">* The field is required</div>
```
[the remark this answers is not in the document]
Comment (tomassychra) [reply packages/web/src/scss/components/TextField/index.html:259]: done

File: packages/web/src/scss/components/TextArea/index.html [hunk 3]
```diff
@@ -70,6 +70,18 @@
     </div>
     <button type="button" class="Button Button--primary Button--medium">Button</button>
   </div>
+
+  <div class="TextArea">
+    <label for="textareaWithHelperText" class="TextArea__label TextArea__label--required">Label</label>
+    <textarea id="textareaWithHelperText" class="TextArea__input" placeholder="Placeholder">Filled</textarea>
+    <div class="TextArea__helperText">* The field is required</div>
```
[the remark this answers is not in the document]
Comment (tomassychra) [reply packages/web/src/scss/components/TextArea/index.html:77]: done

File: packages/web/src/scss/components/TextArea/index.html [hunk 4]
```diff
@@ -70,6 +70,18 @@
...
+  </div>
+  <div class="TextArea TextArea--danger">
+    <label for="textareaWithHelperTextAndValidationMessage" class="TextArea__label TextArea__label--required">Label</label>
+    <textarea id="textareaWithHelperTextAndValidationMessage" class="TextArea__input" placeholder="Placeholder">Filled</textarea>
+    <div class="TextArea__helperText">* The field is required</div>
```
[the remark this answers is not in the document]
Comment (tomassychra) [reply packages/web/src/scss/components/TextArea/index.html:82]: done

File: packages/web-twig/src/Resources/components/TextArea/README.md [hunk 5]
```diff
@@ -4,14 +4,14 @@ This is Twig implementation of the [TextArea] component.
 
 Basic example usage:
 
-```twig
-<TextArea id="example" label="Label" name="example"></TextArea>
+```html
```
[the remark this answers is not in the document]
Comment (tomassychra) [reply packages/web-twig/src/Resources/components/TextArea/README.md]: done

File: packages/web-react/src/components/TextFieldBase/TextFieldBase.tsx [hunk 6]
```diff
@@ -23,6 +23,12 @@ export const TextFieldBase = (props: SpiritTextFieldBaseProps) => {
       propertyName: 'validationState',
     },
   });
+  useDeprecationMessage({
+    method: 'custom',
+    trigger: !!(props?.message && !props?.validationState),
+    componentName: props?.isMultiline ? 'TextArea' : 'TextField',
+    customText: 'Message without validation status will be removed in the next version. Use `helperText` instead.',
```
Comment (literat) [inline packages/web-react/src/components/TextFieldBase/TextFieldBase.tsx:30]:
```suggestion
    customText: 'message prop without validationState prop will be unsupported in the next version. Use `helperText` instead.',
```

What do you think? It sounds better to me like this.

File: packages/web-twig/src/Resources/components/TextArea/README.md [hunk 7]
```diff
@@ -5,13 +5,13 @@ This is Twig implementation of the [TextArea] component.
 Basic example usage:
 
 ```twig
-<TextArea id="example" label="Label" name="example"></TextArea>
+<Textarea id="example" label="Label" name="example"></textarea>
```
Comment (literat) [inline packages/web-twig/src/Resources/components/TextArea/README.md:8]:
```suggestion
<TextArea id="example" label="Label" name="example"></TextArea>
```

File: packages/web-twig/src/Resources/components/TextArea/README.md [hunk 8]
```diff
@@ -5,13 +5,13 @@ This is Twig implementation of the [TextArea] component.
...
 ```
 
 Advanced example usage:
 
 ```twig
-<TextArea
+<Textarea
```
Comment (literat) [inline packages/web-twig/src/Resources/components/TextArea/README.md:14]:
```suggestion
<TextArea
```

File: packages/web-twig/src/Resources/components/TextArea/README.md [hunk 9]
```diff
@@ -21,9 +21,10 @@ Advanced example usage:
   maxlength="180"
   rows="10"
   isRequired
+  helperText="custom helper text"
 >
     TextArea
-</TextArea>
+</textarea>
```
Comment (literat) [inline packages/web-twig/src/Resources/components/TextArea/README.md:27]:
```suggestion
</TextArea>
```

File: packages/web-twig/src/Resources/components/TextFieldBase/TextFieldBase.twig [hunk 10]
```diff
@@ -60,6 +62,9 @@
 {% if _validationState === 'error' %}
     {% deprecated _rootClassName + ': The "error" value for "validationState" property will be removed in the next major version. Use "danger" instead.'️️️ %}
 {% endif %}
+{% if _message and _validationState === null %}
+    {% deprecated _rootClassName + ': Message without validation status will be removed in the next version. Use `helperText` instead.'️️️ %}
```
Comment (literat) [inline packages/web-twig/src/Resources/components/TextFieldBase/TextFieldBase.twig:66]: Same as I suggested in React?

**Review by literat — CHANGES_REQUESTED** [review]
Please review your changes after each modification. TextArea readme examples are still badly written. Please, fix that and it will be ✅ . Other changes are non-blocking suggestions.

File: packages/web-react/src/components/TextFieldBase/TextFieldBase.tsx (hunk 6 above)
Comment (tomassychra) [reply packages/web-react/src/components/TextFieldBase/TextFieldBase.tsx]: fixed

File: packages/web-twig/src/Resources/components/TextArea/README.md (hunk 7 above)
Comment (tomassychra) [reply packages/web-twig/src/Resources/components/TextArea/README.md:8]: fixed

File: packages/web-twig/src/Resources/components/TextArea/README.md (hunk 8 above)
Comment (tomassychra) [reply packages/web-twig/src/Resources/components/TextArea/README.md:14]: fixed

File: packages/web-twig/src/Resources/components/TextArea/README.md (hunk 9 above)
Comment (tomassychra) [reply packages/web-twig/src/Resources/components/TextArea/README.md:27]: fixed

File: packages/web-twig/src/Resources/components/TextFieldBase/TextFieldBase.twig (hunk 10 above)
Comment (tomassychra) [reply packages/web-twig/src/Resources/components/TextFieldBase/TextFieldBase.twig:66]: fixed

**Review by literat — APPROVED** [review]
LGTM 👍

File: packages/web-react/src/components/TextFieldBase/TextFieldBase.tsx [hunk 11]
```diff
@@ -23,6 +23,13 @@ export const TextFieldBase = (props: SpiritTextFieldBaseProps) => {
       propertyName: 'validationState',
     },
   });
+  useDeprecationMessage({
+    method: 'custom',
+    trigger: !!(props?.message && !props?.validationState),
+    componentName: props?.isMultiline ? 'TextArea' : 'TextField',
+    customText:
+      'message prop without validationState prop will be unsupported in the next version. Use `helperText` instead.',
```
Comment (crishpeen) [inline packages/web-react/src/components/TextFieldBase/TextFieldBase.tsx:31]:
```suggestion
      '`message` prop without `validationState` prop will be unsupported in the next version. Use `helperText` instead.',
```

File: packages/web-twig/src/Resources/components/TextFieldBase/TextFieldBase.twig [hunk 12]
```diff
@@ -14,6 +14,7 @@
 {%- set _type = props.type | default('text') -%}
 {%- set _validationState = props.validationState | default(null) -%}
 {%- set _value = props.value | default(null) -%}
+{%- set _helperText = props.helperText | default(null) -%}
```
Comment (crishpeen) [inline packages/web-twig/src/Resources/components/TextFieldBase/TextFieldBase.twig:17]: 🔡 ?

File: packages/web-twig/src/Resources/components/TextFieldBase/TextFieldBase.twig [hunk 13]
```diff
@@ -60,6 +62,9 @@
 {% if _validationState === 'error' %}
     {% deprecated _rootClassName + ': The "error" value for "validationState" property will be removed in the next major version. Use "danger" instead.'️️️ %}
 {% endif %}
+{% if _message and _validationState === null %}
+    {% deprecated _rootClassName + ': message prop without validationState prop will be unsupported in the next version. Use `helperText` instead.'️️️ %}
```
Comment (crishpeen) [inline packages/web-twig/src/Resources/components/TextFieldBase/TextFieldBase.twig:66]:
```suggestion
    {% deprecated _rootClassName + ': `message` prop without `validationState` prop will be unsupported in the next version. Use `helperText` instead.'️️️ %}
```

File: packages/web/src/scss/components/TextArea/index.html [hunk 14]
```diff
@@ -70,6 +70,18 @@
     </div>
     <button type="button" class="Button Button--primary Button--medium">Button</button>
   </div>
+
+  <div class="TextArea">
+    <label for="textareaWithHelperText" class="TextArea__label TextArea__label--required">Label</label>
+    <textarea id="textareaWithHelperText" class="TextArea__input" placeholder="Placeholder">Filled</textarea>
+    <div class="TextArea__helperText">This is helper text</div>
+  </div>
+  <div class="TextArea TextArea--danger">
```
Comment (crishpeen) [inline packages/web/src/scss/components/TextArea/index.html:79]:
```suggestion
  
  <div class="TextArea TextArea--danger">
```
empty line

File: packages/web/src/scss/components/TextField/index.html [hunk 15]
```diff
@@ -252,6 +252,18 @@
     </div>
     <button type="button" class="Button Button--primary Button--medium">Button</button>
   </div>
+
+  <div class="TextField">
+    <label for="textfieldWithHelperText" class="TextField__label TextField__label--required">Label</label>
+    <input type="text" id="textfieldWithHelperText" class="TextField__input" placeholder="Placeholder" />
+    <div class="TextField__helperText">This is helper text</div>
+  </div>
+  <div class="TextField TextField--danger">
```
Comment (crishpeen) [inline packages/web/src/scss/components/TextField/index.html:261]:
```suggestion

  <div class="TextField TextField--danger">
```
empty line

**Review by crishpeen — CHANGES_REQUESTED** [review]
please add twig snapshot tests

File: packages/web-twig/src/Resources/components/TextFieldBase/TextFieldBase.twig (hunk 12 above)
Comment (tomassychra) [reply packages/web-twig/src/Resources/components/TextFieldBase/TextFieldBase.twig:17]: done!

File: packages/web-react/src/components/TextFieldBase/TextFieldBase.tsx (hunk 11 above)
Comment (tomassychra) [reply packages/web-react/src/components/TextFieldBase/TextFieldBase.tsx:31]: done

File: packages/web/src/scss/components/TextArea/index.html (hunk 14 above)
Comment (tomassychra) [reply packages/web/src/scss/components/TextArea/index.html:79]: done

File: packages/web/src/scss/components/TextField/index.html (hunk 15 above)
Comment (tomassychra) [reply packages/web/src/scss/components/TextField/index.html:261]: done

File: packages/web-react/src/components/TextArea/__tests__/TextArea.test.tsx [hunk 16]
```diff
@@ -45,12 +45,19 @@ describe('TextArea', () => {
   });
 
   it('should have message', () => {
-    const dom = render(<TextArea id="textarea" label="Label" message="text" />);
+    const dom = render(<TextArea id="textarea" label="Label" message="text" validationState="danger" />);
```
Comment (crishpeen) [inline packages/web-react/src/components/TextArea/__tests__/TextArea.test.tsx:48]: maybe we should test it without `validationState` too?

File: packages/web-twig/src/Resources/components/TextFieldBase/TextFieldBase.twig [hunk 17]
```diff
@@ -124,6 +129,11 @@
             />
         {% endif %}
     {% endif %}
+    {% if _helperText %}
+        <span class="{{ _helperTextClassName }}">
```
Comment (crishpeen) [inline packages/web-twig/src/Resources/components/TextFieldBase/TextFieldBase.twig:133]:
```suggestion
        <div class="{{ _helperTextClassName }}">
```
in `web` and `web-react` we use `div`. Let's unify it. Please update snapshots as well.

**Review by crishpeen — APPROVED** [review]

File: packages/web-react/src/components/TextArea/__tests__/TextArea.test.tsx (hunk 16 above)
Comment (tomassychra) [reply packages/web-react/src/components/TextArea/__tests__/TextArea.test.tsx:48]: ok

File: packages/web-twig/src/Resources/components/TextFieldBase/TextFieldBase.twig (hunk 17 above)
Comment (tomassychra) [reply packages/web-twig/src/Resources/components/TextFieldBase/TextFieldBase.twig]: fixed

**Pull request merged**